demos/app/iot_demo_arguments.c

```diff
@@ -133,22 +143,37 @@ static bool _validateArguments( const IotDemoArguments_t * pArguments )
             IOT_SET_AND_GOTO_CLEANUP( false );
         }
 
-        /* Check that a client certificate path was set. */
-        if( ( pArguments->pClientCertPath == NULL ) ||
-            ( strlen( pArguments->pClientCertPath ) == 0 ) )
+        /* If the host is connecting to the MQTT broker hosted by AWS IoT Core,
+         * there must either be a set of X.509 credentials or a 
+         * username/password. Therefore, check that here in order to facilitate 
+         * debugging. For other MQTT brokers, assume that the CLI arguments are
+         * as intended.
+         */
+        if( pArguments->awsIotMqttMode == true )
         {
-            IotLogError( "Client certificate path not set. Exiting." );
+            if( ( pArguments->pUserName == NULL ) ||
+                ( strlen( pArguments->pUserName ) == 0 ) ||
+                ( pArguments->pPassword == NULL ) ||
+                ( strlen( pArguments->pPassword ) == 0 ) )
+            {
+                /* Check that a client certificate path was set. */
+                if( ( pArguments->pClientCertPath == NULL ) ||
+                    ( strlen( pArguments->pClientCertPath ) == 0 ) )
+                {
+                    IotLogError( "Either username/password or client certificate path must be set. Exiting." );
 
-            IOT_SET_AND_GOTO_CLEANUP( false );
-        }
+                    IOT_SET_AND_GOTO_CLEANUP( false );
+                }
 
-        /* Check that a client certificate private key was set. */
-        if( ( pArguments->pPrivateKeyPath == NULL ) ||
-            ( strlen( pArguments->pPrivateKeyPath ) == 0 ) )
-        {
-            IotLogError( "Client certificate private key not set. Exiting." );
+                /* Check that a client certificate private key was set. */
+                if( ( pArguments->pPrivateKeyPath == NULL ) ||
+                    ( strlen( pArguments->pPrivateKeyPath ) == 0 ) )
+                {
+                    IotLogError( "Either username/password or private key path must be set. Exiting." );
 
-            IOT_SET_AND_GOTO_CLEANUP( false );
+                    IOT_SET_AND_GOTO_CLEANUP( false );
+                }
```

Comment:
Is a valid `pArguments->pRootCaPath` also required to connect to AWS IoT with a client certificate/private key?

Reply:
Yes, the proposed validation rules for the CLI options are that "secure mode" requires an explicit root certificate for validating the server chain. Beyond that, if you're also in "AWS IoT mode," then a client credential is also required: username/password or certificate/private key.